Added test cases for service call without sidecar

yu2003w opened the pull request:
Signed-off-by: Wei Yu(Jared) <yuweiw@cn.ibm.com>

<!--
Request Prow to automatically lint any go code in this PR:

/lint
-->

Fixes #4028

## Proposed Changes
Following tests added:
* Service call scale from zero without side car
* Service A without side car call service B with side car
* Service A with side car call Service B without side car

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
None
```

mattmoor-sockpuppet commented on third_party/istio-1.2-latest:
```diff
@@ -0,0 +1 @@
+istio-1.2.0
```
<!--whitespace-bot-->
```suggestion
istio-1.2.0

```

tcnghia commented:
/ok-to-test

tcnghia commented:
/assign

tcnghia commented on test/e2e/service_to_service_test.go:
```diff
@@ -216,5 +221,145 @@ func TestServiceToServiceCallFromZero(t *testing.T) {
+}
+
+// set up tests with sidecar disabled for both service
+func TestSvcToSvcCallFromZeroWithoutSidecar(t *testing.T) {
```
can we extract this test into a common function and create a table test to call them all?  some examples:

https://github.com/knative/serving/blob/master/test/e2e/service_to_service_test.go#L177
https://github.com/knative/serving/blob/master/test/e2e/grpc_test.go#L204

yu2003w replied:
Thanks for the comments. Updated code.

tcnghia commented:
/test pull-knative-serving-integration-tests

yt3liu commented:
/cc @ZhiminXiang

markusthoemmes commented on test/e2e/service_to_service_test.go:
```diff
@@ -216,5 +234,57 @@ func TestServiceToServiceCallFromZero(t *testing.T) {
+
+// service call to servie call with istio sidecar injected automatically
+func TestSvcToSvcCallBothInject(t *testing.T) {
+
```
Remove the empty line.

markusthoemmes commented on test/e2e/service_to_service_test.go:
```diff
@@ -70,7 +70,7 @@ func sendRequest(t *testing.T, clients *test.Clients, resolvableDomain bool, dom
 }
 
-func testProxyToHelloworld(t *testing.T, clients *test.Clients, helloworldDomain string) {
+func testProxyToHelloworld(t *testing.T, clients *test.Clients, helloworldDomain string, revAnno *map[string]string) {
```
Why pass this as a pointer? Maps are pointer types anyway.

markusthoemmes commented on test/e2e/service_to_service_test.go:
```diff
@@ -138,12 +143,9 @@ func testProxyToHelloworld(t *testing.T, clients *test.Clients, helloworldDomain
-	defer cancel()
-
-	clients := Setup(t)
+// issue #3620 added the possibility to opt-out of the mesh on the revision pod side.
```
I'd remove this line, it feels redundant.

markusthoemmes commented on test/e2e/service_to_service_test.go:
```diff
@@ -174,22 +184,22 @@ func TestServiceToServiceCall(t *testing.T) {
-
-	clients := Setup(t)
+// injectA indicates whether or not inject istio sidecar into httpproxy service
+// injectB indicates whether or not inject istio sidecar into helloworld service
```
Use the same comment above, it's more comprehensive imo.

markusthoemmes commented on test/e2e/service_to_service_test.go:
```diff
@@ -204,7 +214,15 @@ func TestServiceToServiceCallFromZero(t *testing.T) {
+	opts := v1a1test.Options{}
+	if !injectA {
+		opts.RevisionTemplateAnnotations = revAnno
+	}
```
Why is this the other way round as above? (above has !injectB here).

markusthoemmes commented on test/e2e/service_to_service_test.go:
```diff
@@ -174,22 +184,22 @@ func TestServiceToServiceCall(t *testing.T) {
+	if injectA {
+		revAnno = nil
+	}
 	for _, tc := range testCases {
```
Since we have a subtest setup anyway:

Should we ditch `testServiceToServiceCallFromZero` altogether and add an option to the test-table saying whether we want to wait for scale to 0 or not? That should cut down on code duplication in this test I think.

yu2003w commented:
@markusthoemmes I updated tests and did not touch parts for short names.
I rebased the code due to conflicts. 
Could you please help to review again?

markusthoemmes commented on test/e2e/service_to_service_test.go:
```diff
@@ -219,7 +229,11 @@ func TestServiceToServiceCallViaActivator(t *testing.T) {
+		&v1a1test.Options{
+			RevisionTemplateAnnotations: map[string]string{
+				"sidecar.istio.io/inject": strconv.FormatBool(injectB),
+			},
```
Can you move this into the `WithConfigAnnotations` helper below? I think it's kinda redundant otherwise.

yu2003w replied:
Updated. I missed the fact that WithConfigAnnotations could do the same thing.
Thanks.

yu2003w commented:
/retest

yu2003w commented:
> /approve
> 
> The changes LGTM. I'd like @tcnghia to sign off and be aware though.
> 
> @yu2003w have you done the MESH=1 test run @tcnghia suggested above?

I ran the tests in my dev environment and tests passed with ISTIO_MESH=1

tcnghia commented:
/lgtm
/approve

thanks a lot!
